File: odometer/src/odometer.cpp
#include "odometer.h"
#include "ros/ros.h"


odo::odo(ros::NodeHandle *nh){
    speed=0;
    g_count=0;
    total_count=0;
    speed_pub=nh->advertise<odometer::speed_msg>("/speed",10);
    timer=nh->createTimer(ros::Duration(0.5),&odo::publish_speed,this);
}

double odo::calculate_speed(){
    total_count+=g_count; //refresh odom
    speed=(g_count/PPR/ratio)*CIRCUL*3.6*2; //count to KPH
    g_count=0;
    return speed;
}

void odo::publish_speed(const ros::TimerEvent& ){
    odometer::speed_msg msg;
    msg.kph=calculate_speed();
    msg.odo=total_count/PPR/ratio*CIRCUL;
    speed_pub.publish(msg);
}
void count_up(){
    g_count++;
}

int main(int argc, char**argv){
    wiringPiSetupPhys();
    pinMode(SENSOR_PIN,INPUT);
    ros::init(argc,argv,"odo_node");
    ros::NodeHandle nh;
    odo odometer(&nh);
    wiringPiISR(SENSOR_PIN,INT_EDGE_BOTH,&count_up); //sensing
    ros::spin();
}

File: odometer/src/odometer.py
#!/usr/bin/env python

import rospy
from std_msgs.msg import TimerEvent
from odometer.msg import speed_msg
import RPi.GPIO as GPIO

SENSOR_PIN = 23
PPR = 32
g_count = 0

class Odometer:
    def __init__(self, nh):
        self.speed = 0.0
        self.total_count = 0
        self.speed_pub = nh.Publisher('/speed', speed_msg, queue_size=10)
        self.timer = rospy.Timer(rospy.Duration(0.5), self.publish_speed)

    def calculate_speed(self):
        global g_count
        self.total_count += g_count
        self.speed = (g_count * 1.527 / PPR) * 3.6 * 2  # Count to KPH
        g_count = 0
        rospy.loginfo("KPH: %f km/h", self.speed)
        return self.speed

    def publish_speed(self, event):
        msg = speed_msg()
        msg.kph = self.calculate_speed()
        msg.odo = self.total_count / PPR * 1.527
        self.speed_pub.publish(msg)

def count_up(channel):
    global g_count
    g_count += 1

def main():
    GPIO.setmode(GPIO.BCM)
    GPIO.setup(SENSOR_PIN, GPIO.IN)
    rospy.init_node('odo_node')
    nh = rospy.NodeHandle()
    odometer = Odometer(nh)
    GPIO.add_event_detect(SENSOR_PIN, GPIO.BOTH, callback=count_up)
    rospy.spin()

if __name__ == '__main__':
    main()
